### ISSUE TITLE
False positive wrong_self_convention inside trait impls

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#6983]
Beginning with the most recent nightly (clippy 0.1.53 5e65467 2021-03-26, rustc 1.53.0-nightly 5e65467ef 2021-03-26) the following code incorrectly triggers `wrong_self_convention`. The trait method signature is controlled by the trait; it doesn't matter that the impl is for a type implementing `Copy`, that doesn't mean that the trait can or should take self by value. In particular there are likely other impls which are not Copy.

```rust
pub struct Thing;

pub trait Trait {
    fn to_thing(&self) -> Thing;
}

impl Trait for u8 {
    fn to_thing(&self) -> Thing {
        Thing
    }
}
```

```console
warning: methods with the following characteristics: (`to_*` and `self` type is `Copy`) usually take `self` by value
 --> src/main.rs:8:17
  |
8 |     fn to_thing(&self) -> Thing {
  |                 ^^^^^
  |
  = note: `#[warn(clippy::wrong_self_convention)]` on by default
  = help: consider choosing a less ambiguous name
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention
```

Mentioning @mgacek8 @llogiq who touched this lint recently in #6924.

**Comment by dtolnay** [comment]
Mentioning @matthiaskrgr too since I am surprised this was not caught by b9a7a2a2750ee112f71d95b3f588bc9d51e949a5 -- all three of the above affected crates (quote, syn, cxx) are in lintcheck_crates.

https://github.com/rust-lang/rust-clippy/blob/6f2a6fe84fc811c04b5693531042deb44b077233/lintcheck/lintcheck_crates.toml#L17

https://github.com/rust-lang/rust-clippy/blob/6f2a6fe84fc811c04b5693531042deb44b077233/lintcheck/lintcheck_crates.toml#L21

https://github.com/rust-lang/rust-clippy/blob/6f2a6fe84fc811c04b5693531042deb44b077233/lintcheck/lintcheck_crates.toml#L26

**Comment by llogiq** [comment]
Oops. I was under the impression that methods taking `&self` should be named `as_*` instead of `to_`.

**Comment by extrawurst** [comment]
this is even more problematic because the trait `ToString` in `std` expects me to pass `self` as `&self` in `to_string` and now clippy starts complaining about that :)
